### PR TITLE
[Issue #2770] replace class based fetching with endpoint configuration

### LINKED ISSUE #2770: Next is making too many requests
### Summary

@chouinar noticed that
* for each search page load, 4 search requests are being received by the API, and 
* for each opportunity listing page load, 2 opportunity listing requests are being received

The expectation is that only one request is made in each case.

Initial research points to Next's request memoization not working as expected.

### Reproducibility

I tried multiple times and saw the bug every time

### How to reproduce the bug

1. visit a search page
2. tail your docker logs and notice that 4 requests are received
3. visit an opportunity page
4. tail your docker logs and notice that 2 requests are received

### Browser

None

### Code of Conduct

- [x] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by doug-s-nava**
## Summary
Fixes #2770

Note that https://github.com/HHS/simpler-grants-gov/pull/2787 also solves this issue, we only need to take one or the other

## Changes proposed
So this builds on https://github.com/HHS/simpler-grants-gov/pull/2787, and largely accomplishes the same goal, while also completely refactoring our fetching system.

This is a spec PR and I'm not expecting it to be adopted without some discussion.

## Context for reviewers
When debugging the memoization issues for ticket 2770 it became difficult for me in the current system to determine if our classical API fetching system was causing problems for Next in deduplicating requests, and when implementing a React cache call on the existing system, it was somewhat complicated to isolate the request functionality in order to wrap things correctly.

I started to build out a simpler system for debugging only with some success (though I was never able to fully debug the situation for other reasons).

In https://github.com/HHS/simpler-grants-gov/pull/2787, without a big refactor, I made a change to isolate the functions from the two fetch APIs that are actually being called to make the requests, which made me realize that only these functions need to be exposed from the API classes they live in. In this case, it seems a little overdone to be using classes for this purpose, so I went ahead and completed the POC for the new endpoint config based system I was working off of.

### Test steps

1. visit an opportunity detail on main while tailing API docker logs `docker logs grants-api -f`
2. _VERIFY_: two separate "start request" logs are visible
1. visit an opportunity detail on this branch while tailing API docker logs `docker logs grants-api -f`
2. _VERIFY_: one "start request" log is visible

### CODE REVIEW

File: frontend/tests/api/SearchFetcher.test.ts [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+
+    expect(result).toEqual({
+      actionType: "fun",
+      data: {},
+      fieldChanged: "baseball",
+    });
+  });
+});
+
+describe("buildFilters", () => {
```
Comment (doug-s-nava) [inline frontend/tests/api/SearchFetcher.test.ts:70]: the rests of these tests aren't new

File: frontend/tests/api/FetchHelpers.test.ts [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+  .fn()
+  .mockResolvedValue({ data: [], errors: [], warnings: [] });
+
+const fetchMock = jest.fn().mockResolvedValue({
+  json: responseJsonMock,
+  ok: true,
+  status: 200,
+});
+
+describe("createRequestUrl", () => {
```
Comment (doug-s-nava) [inline frontend/tests/api/FetchHelpers.test.ts:28]: this isn't new

File: frontend/src/app/api/searchFetcher.ts [hunk 3]
```diff
@@ -35,43 +34,32 @@ const filterNameMap = {
...
-export default class SearchOpportunityAPI extends BaseApi {
-  get namespace(): string {
-    return "opportunities";
-  }
-
-  async searchOpportunities(searchInputs: QueryParamData) {
-    const { query } = searchInputs;
-    const filters = buildFilters(searchInputs);
-    const pagination = buildPagination(searchInputs);
+export const searchForOpportunities = async (searchInputs: QueryParamData) => {
```
Comment (doug-s-nava) [inline frontend/src/app/api/searchFetcher.ts:37]: kept the necessary custom search related functionality from the SearchOpportunityApi here

File: frontend/src/app/api/fetchers.ts [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+} from "src/app/api/fetchHelpers";
+import { APIResponse } from "src/types/apiResponseTypes";
+import { OpportunityApiResponse } from "src/types/opportunity/opportunityResponseTypes";
+import { QueryParamData } from "src/types/search/searchRequestTypes";
+import { SearchAPIResponse } from "src/types/search/searchResponseTypes";
+
+import { cache } from "react";
+
+// returns a function which can be used to make a request to an endpoint defined in the passed config
+export function requesterForEndpoint<ResponseType extends APIResponse>({
```
Comment (doug-s-nava) [inline frontend/src/app/api/fetchers.ts:24]: originally had this living elsewhere, but testing was much easier if I isolated it away from the other helper fns

File: frontend/src/app/api/fetcherHelpers.ts [hunk 5]
```diff
@@ -26,95 +26,41 @@ export interface HeadersDict {
...
 }
 
-export default abstract class BaseApi {
-  // Root path of API resource without leading slash, can be overridden by implementing API classes as necessary
-  get basePath() {
-    return environment.API_URL;
-  }
+// Configuration of headers to send with all requests
+// Can include feature flags in child classes
+export function getDefaultHeaders(): HeadersDict {
```
Comment (doug-s-nava) [inline frontend/src/app/api/fetcherHelpers.ts:31]: split most of the old BaseAPI functionality into individual functions

File: frontend/src/app/api/endpointConfigs.ts [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+
+export interface EndpointConfig {
+  basePath: string;
+  version: string;
+  namespace: string;
+  method: ApiMethod;
+}
+
+export const opportunitySearchEndpoint = {
+  basePath: environment.API_URL,
```
Comment (doug-s-nava) [inline frontend/src/app/api/endpointConfigs.ts:12]: would like to structure this so basePath could be defaulted, but didn't go that far today

File: frontend/src/app/api/fetchers.ts [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+    return response;
+  };
+}
+
+export const fetchOpportunity = cache(
+  requesterForEndpoint<OpportunityApiResponse>(fetchOpportunityEndpoint),
+);
+
+export const requestOpportunitySearch = requesterForEndpoint<SearchAPIResponse>(
```
Comment (acouch) [inline frontend/src/app/api/fetchers.ts:70]: [nit]: Is there a difference between "request" and "fetch," ie GET vs POST? If not would suggest a consistent name.

File: frontend/src/app/api/searchFetcher.ts [hunk 8]
```diff
@@ -35,43 +34,32 @@ const filterNameMap = {
...
+    requestBody.query = query;
+  }
 
-    const response = await this.request<SearchAPIResponse>(
-      "POST",
-      "search",
-      searchInputs,
-      requestBody,
-    );
+  const response = await requestOpportunitySearch("", { body: requestBody });
```
Comment (acouch) [inline frontend/src/app/api/searchFetcher.ts:53]: [nit]: Making `subpath` optional would make this a little cleaner.

**Review by acouch — APPROVED** [review]
🎉 awesome work. New interface looks easier to work with and extend.

**Review by mdragon — DISMISSED** [review]
Looks good, great to get this addressed for our scaling.

**Review by mdragon — APPROVED** [review]

**Pull request merged**